analyzer/javacg-opal/src/main/java/eu/fasten/analyzer/javacgopal/data/analysis/OPALClassHierarchy.java

```diff
@@ -234,17 +262,18 @@ public Map<Object, Object> getInternalMetadata(final Map<List<Integer>, Map<Obje
     }
 
     /**
-     * Append a sub-graph to already existing PartialJavaCallGraph.
-     *  @param source      source method
-     * @param targets     list of targets
-     * @param resultGraph already existing PartialJavaCallGraph
+     * Append a sub-graph to already existing PartialJavaCallGraph. It is thread-safe.
+     *
+     * @param source       source method
+     * @param targets      list of targets
+     * @param resultGraph  already existing PartialJavaCallGraph
      * @param callSiteOnly
      */
-    public void appendGraph(final Object source,
-                            final Iterator<Tuple2<Object, Iterator<DeclaredMethod>>> targets,
-                            final Stmt<DUVar<ValueInformation>>[] stmts,
-                            final JavaGraph resultGraph, List<Integer> incompeletes,
-                            final Set<Integer> visitedPCs, CallPreservationStrategy callSiteOnly) {
+    public synchronized void appendGraph(final Object source,
```

Comment:
This synchronized method is not the most efficient way of parallelizing this part. I suggest that we make the data structure thread-safe. Data structures that we need to change would be OpalClassHierarchy and JavaGraph. Perhaps some ConcurrentMap AtomicInt would do the job. However, this is a bit complicated we may need to debug it together if needed.

Reply:
I agree. The most efficient solution should also make the data structures related to CGs thread-safe. For now, to pass tests, I've added `synchronized` to `appendGraph()`. 
It would be nice to work together and make `appendGraph()` more efficient for multi-threading.